src/TokenParser/SpacelessTokenParser.php

```diff
@@ -30,13 +33,23 @@ class SpacelessTokenParser extends AbstractTokenParser
 {
     public function parse(Token $token)
     {
-        $lineno = $token->getLine();
+        $this->parser->getStream()->injectTokens([
+            new Token(Token::NAME_TYPE, 'spaceless', $token->getLine()),
+        ]);
 
+        $name = $this->parser->getVarName();
+        $ref = new BlockReferenceExpression(new ConstantExpression($name, $token->getLine()), null, $token->getLine(), $this->getTag());
+
+        $filter = $this->parser->getExpressionParser()->parseFilterExpressionRaw($ref, $this->getTag());
         $this->parser->getStream()->expect(Token::BLOCK_END_TYPE);
+
         $body = $this->parser->subparse([$this, 'decideSpacelessEnd'], true);
         $this->parser->getStream()->expect(Token::BLOCK_END_TYPE);
 
-        return new SpacelessNode($body, $lineno, $this->getTag());
+        $block = new BlockNode($name, $body, $token->getLine());
```

Comment:
Using a block node here is a BC break: a block creates its own isolated block, while `{% spaceless %}` is not a scoping element.

Reply:
This part has been reverted since the merge.